pyro/distributions/empirical.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import absolute_import, division, print_function
+
+import numbers
+
+import math
+import torch
+
+from pyro.distributions.torch_distribution import TorchDistribution
+from pyro.distributions.torch import Categorical
+from pyro.distributions.util import copy_docs_from, is_validation_enabled
+from pyro.distributions.util import log_sum_exp
+
+
+@copy_docs_from(TorchDistribution)
+class Empirical(TorchDistribution):
+    r"""
+    Empirical distribution associated with the sampled data.
+    """
+
+    arg_constraints = {}
+
+    def __init__(self):
+        self._samples = None
+        self._log_weights = None
+        self._categorical = None
+        self._samples_buffer = []
+        self._weights_buffer = []
+        super(TorchDistribution, self).__init__()
+
+    def _append_from_buffer(self, tensor, buffer):
+        """
+        Append values from the buffer to the finalized tensor, along the
+        leftmost dimension.
+
+        :param torch.Tensor tensor: tensor containing existing values.
+        :param list buffer: list of new values.
+        :return: tensor with new values appended at the bottom.
+        """
+        buffer_tensor = torch.stack(buffer, dim=0)
+        return torch.cat([tensor, buffer_tensor], dim=0)
+
+    def _finalize(self):
+        """
+        Appends values collected in the samples/weights buffers to their
+        corresponding tensors.
+        """
+        if not self._samples_buffer:
+            return
+        self._samples = self._append_from_buffer(self._samples, self._samples_buffer)
+        self._log_weights = self._append_from_buffer(self._log_weights, self._weights_buffer)
+        self._categorical = Categorical(logits=self._log_weights)
+        # Reset buffers.
+        self._samples_buffer, self._weights_buffer = [], []
+
+    @property
+    def sample_size(self):
+        """
+        Number of samples that constitute the empirical distribution.
+
+        :return int: number of samples collected.
+        """
+        self._finalize()
+        if self._samples is None:
+            return 0
+        return self._samples.size(0)
+
+    def add(self, value, log_weight=None, weight=None):
```

Comment:
nit: `weight` seems more standard than `log_weight`, so I'd be less surprised if if was to the left of `log_weight`

Reply:
Will change.